### DISCUSSION TITLE
Consider integrations of NTM reference genomes

### DISCUSSION THREAD

**Discussion by abhi18av** [mycobactopia-org/MTBseq-nf#68] [General]
@Mxrcon here's a dumping ground of ideas regarding this investigation.

**Comment by Mxrcon** [comment]
# CONTEXT:
During our discussions on slack, Hanno was curious to see how mtbseq would interact with the usage of other references. Preferably _M. bovis_ and _M. canetii_ and see if is there any difference from the standard reference: _M. tuberculosis h37rv_

**Reply by Mxrcon** [reply]
[MTBSEQ manual about using custom refs](https://github.com/ngs-fzb/MTBseq_source/blob/master/MANUAL.md#exchanging-the-reference-genome)

Decided to add multiple references to my conda env with mtbseq.
    
- Standard references:
  - Mycobacterium tuberculosis H37rv (M._tuberculosis_H37Rv_2015-11-13)
  - Mycobacterium abscessus (M._abscessus_CIP-104536T_2014-02-03)
  - Mycobacterium chimaera (M._chimaera_DSM44623_2016-01-28)
  - Mycobacterium fortuitum (M._fortuitum_CT6_2016-01-08)

Every ref has 2 main files: `.fasta` and `genes.txt`

- NOTE: _M. tuberculosis_ has additional files: 
  - `.fasta.bwt` 
  - `.dict`
  - `.fasta.fai`
  - `.fasta.pac`
  - `.fasta.amb`
  - `.fasta.sa`
  - `.fasta.ann` 

- REF seqs used:
  - [_M. canetii_](https://www.ncbi.nlm.nih.gov/nuccore/NC_015848.1) 
  - [_M. bovis_](https://www.ncbi.nlm.nih.gov/nuccore/NC_012207.1)

**Comment by Mxrcon** [comment]
After some hard work, i came up with a python script to generate the `genes.txt` file, i used the M. Tuberculosis H37rv gbk from ncbi and tried to replicate exactly the output.

[script link](https://github.com/Mxrcon/create_mtbseq_ref/blob/main/create_mtbseq_ref.py)

I double checked the outputs using the mtbseq source `genes.txt` and the outputs from my script and running git diff between them, I could replicate exactly the same file so the git diff output was none.

**Reply by abhi18av** [reply]
CC @jloubser

**Reply by jloubser** [reply]
@Mxrcon this is fantastic! appreciate the trouble. I've tested the script a couple of times with M.bovis (LT708304), but with no luck. Just for good measure I tried it with H37Rv and worked perfectly. Then I realised I used the genome assembly instead of the complete sequence record. Tried it again, this time with NC_002945. Exactly the same error. 

I noticed the Genbank record for H37Rv is linear DNA and for LT708304 is circular DNA and thought it might be the issue. It's not. 
I chose this one as this is the reference used by all M.bovis papers (correct name: AF212297).
But I used the BCG reference genome you used and it worked again.
I'll check further what the differences are, but I find this odd.
Any idea why it wouldn't work with these reference sequences?


The error I got with both tries (LT708304 & NC_002945):
<(base) [jloubser@khaos data]$ python3 create_mtbseq_ref.py bovis.gb
Traceback (most recent call last):
  File "/home/jloubser/mtbseq-nf/data/create_mtbseq_ref.py", line 40, in <module>
    create_fasta(output_name, sample_sequence)
  File "/home/jloubser/mtbseq-nf/data/create_mtbseq_ref.py", line 13, in create_fasta
    with open(output_name+".fasta", "w") as output_fasta:
FileNotFoundError: [Errno 2] No such file or directory: 'Mycobacterium_tuberculosis_variant_bovis_AF2122/9723-JUN-2022.fasta'>

**Reply by Mxrcon** [reply]
Hello @jloubser, sorry for my delay. This issue seems to be related to the `/` character on the file name, would you be able to rename the file name to something like: `Mycobacterium_tuberculosis_variant_bovis_AF2122_9723-JUN-2022` and test it again? this should fix the problem

**Reply by jloubser** [reply]
> Hello @jloubser, sorry for my delay. This issue seems to be related to the `/` character on the file name, would you be able to rename the file name to something like: `Mycobacterium_tuberculosis_variant_bovis_AF2122_9723-JUN-2022` and test it again? this should fix the problem

@Mxrcon no problem, you have more pressing things to focus on. I appreciate the time you've put in for this. I'll give this a try.
